src/components/implementation/no_interface/cFE_booter/osfilesys.c

```diff
@@ -1,280 +1,886 @@
+    return 0;
 }
 
-/*
- * Copies information from a tar file header to a fsobj
- */
-uint32 tar_cphdr(uint32 tar_offset, struct fsobj *file)
+int32 file_open(char *path, permission_t permission)
+{
+    assert(openfs);
+    if (!openfs->root) return OS_FS_ERR_PATH_TOO_LONG;
+    if (!path) return OS_FS_ERR_INVALID_POINTER;
+
+    // find the file
+    struct fsobj *file = file_find(path);
+    if (!file) return OS_FS_ERR_PATH_INVALID;
+    if (file->type != FSOBJ_FILE) return OS_FS_ERROR;
+
+    // get a new fd
+    int32 FD = newfd_get(file->ino);
+    if (FD == OS_FS_ERROR) {
+        return OS_FS_ERR_NO_FREE_FDS;
+    }
+
+    struct fd *filedes = &fd_tbl[FD];
+
+    filedes->access                         = permission;
+    filedes->file                           = file;
+    filedes->position.file_pos.open_part    = file->file_part;
+    filedes->position.file_pos.file_offset  = 0;
+    filedes->position.file_pos.part_offset  = 0;
+    file->refcnt++;
+    return FD;
+}
+
+int32 file_close(int32 FD)
 {
-    //TODO: store filenames without paths
-    assert(tar_offset < size);
-    assert(file->filedes > 0);
+    int32 ret = chk_fd(FD);
+    if (ret != OS_FS_SUCCESS) return OS_FS_ERROR;
+
+    struct fd *filedes = &fd_tbl[FD];
+    if (filedes->ino == 0) return OS_FS_ERROR;
 
-    char *location = start;
-    location += tar_offset;
-    file->name          = location;
-    if (*(location + strlen(location) - 1) == '/') {
-        file->type      = FSOBJ_DIR;
-        file->size      = 0;
+    uint32 index = filedes->ino - 1;
+    assert(files[index].refcnt > 0);
+    files[index].refcnt--;
+    filedes->ino = 0;
+    return OS_FS_SUCCESS;
+}
+
+int32 file_close_by_name(char *path)
+{
+    struct fsobj *file = file_find(path);
+    int i;
+
+    for (i = 0 ; i < OS_MAX_NUM_OPEN_FILES + 1 ; i++) {
+        if (fd_tbl[i].file == file){
+            assert(fd_tbl[i].ino = file->ino);
+            file_close(i);
+            return OS_FS_SUCCESS;
+        }
     }
-    else{
-        file->type      = FSOBJ_FILE;
-        file->size      = oct_to_dec(location + 124);
-        file->data      = location + 500;
+    return OS_FS_ERROR;
+}
+
+// converts from the cFE defined permission constants to internal permission type
+// unknown permissions return NONE, cFE should treat none as an error
+permission_t permission_cFE_to_COS(uint32 permission) {
+    switch (permission) {
+        case OS_READ_WRITE :
+            return READ_WRITE;
+        case OS_WRITE_ONLY :
+            return WRITE;
+        case OS_READ_ONLY :
+            return READ;
+        default :
+            return NONE;
+    }
+    PANIC("unreachable statement");
+    return 0;
+}
+
+uint32 permission_cos_to_cFE(permission_t permission) {
+    switch (permission) {
+        case READ_WRITE :
+            return OS_READ_WRITE;
+        case WRITE :
+            return OS_WRITE_ONLY;
+        case READ :
+            return OS_READ_ONLY;
+        default :
```

Comment:
the cFE does not provide any permission which maps to NONE.  Data corruption or something else screwy is going on if a file it has access to is not read, write, or read & write